osu.Game/Rulesets/UI/HitObjectContainer.cs

```diff
@@ -1,78 +1,174 @@
+
+        /// <summary>
+        /// Invoked when a <see cref="DrawableHitObject"/> judgement is reverted.
+        /// </summary>
+        public event Action<DrawableHitObject, JudgementResult> RevertResult;
+
+        private readonly Dictionary<DrawableHitObject, IBindable> startTimeMap = new Dictionary<DrawableHitObject, IBindable>();
+        private readonly Dictionary<HitObjectLifetimeEntry, DrawableHitObject> drawableMap = new Dictionary<HitObjectLifetimeEntry, DrawableHitObject>();
+        private readonly LifetimeEntryManager lifetimeManager = new LifetimeEntryManager();
+
+        [Resolved(CanBeNull = true)]
+        private DrawableRuleset drawableRuleset { get; set; }
 
         public HitObjectContainer()
         {
             RelativeSizeAxes = Axes.Both;
+
+            lifetimeManager.EntryBecameAlive += entryBecameAlive;
+            lifetimeManager.EntryBecameDead += entryBecameDead;
         }
 
-        public virtual void Add(DrawableHitObject hitObject)
+        #region Pooling support
+
+        public void Add(HitObjectLifetimeEntry entry) => lifetimeManager.AddEntry(entry);
+
+        public bool Remove(HitObjectLifetimeEntry entry) => lifetimeManager.RemoveEntry(entry);
+
+        private void entryBecameAlive(LifetimeEntry entry) => addDrawable((HitObjectLifetimeEntry)entry);
+
+        private void entryBecameDead(LifetimeEntry entry) => removeDrawable((HitObjectLifetimeEntry)entry);
+
+        private void addDrawable(HitObjectLifetimeEntry entry)
         {
-            // Added first for the comparer to remain ordered during AddInternal
-            startTimeMap[hitObject] = (hitObject.HitObject.StartTimeBindable.GetBoundCopy(), hitObject.HitObject.StartTime);
-            startTimeMap[hitObject].bindable.BindValueChanged(_ => onStartTimeChanged(hitObject));
+            Debug.Assert(!drawableMap.ContainsKey(entry));
+
+            var drawable = drawableRuleset.GetDrawableRepresentation(entry.HitObject);
+            drawable.OnNewResult += onNewResult;
+            drawable.OnRevertResult += onRevertResult;
+
+            bindStartTime(drawable);
+            AddInternal(drawableMap[entry] = drawable, false);
+        }
+
+        private void removeDrawable(HitObjectLifetimeEntry entry)
+        {
+            Debug.Assert(drawableMap.ContainsKey(entry));
+
+            var drawable = drawableMap[entry];
+            drawable.OnNewResult -= onNewResult;
+            drawable.OnRevertResult -= onRevertResult;
+            drawable.OnKilled();
+
+            drawableMap.Remove(entry);
 
+            unbindStartTime(drawable);
+            RemoveInternal(drawable);
+        }
+
+        #endregion
+
+        #region Non-pooling support
+
+        public virtual void Add(DrawableHitObject hitObject)
+        {
+            bindStartTime(hitObject);
             AddInternal(hitObject);
+
+            hitObject.OnNewResult += onNewResult;
+            hitObject.OnRevertResult += onRevertResult;
```

Comment:
Probably relatively minor, but is the ordering difference compared to the pooling path meaningful in any capacity? (pooling binds first then adds, non-pooling adds first then binds). Probably would read better to unify one way or another if possible.

Reply:
Not meaningful at all, but I'll move the `AddInternal()` to the bottom.